### ISSUE TITLE
Accessing Pin.Callout leads into endless loop with latest Xamarin.Forms

### ISSUE THREAD

**Issue by mmichelbb** [Mapsui/Mapsui#606]
I have compiled Mapsui.Forms against the latest version of Xamarin.Forms. 
Whenever I tried accessing the Callout on a pin (doing it the way as described in Forms.Shared PinSample), my app got stuck in an infinity loop and therefore doing nothing anymore.

I suspect that this is caused by line 457 in Mapsui.UI.Forms.MapView.cs:
`while (_callout == null) ;`

For now my solution is to create and populate the callout manually using reflection and injecting it into Pin._callout afterwards, so that MapView._callout is never called in the first place.

Here is the code I am using to populate the pins:

`
		private Pin CreatePin(MapView mapView, int pinNumber, string label, string address, double latitude, double longitude, IconTypes iconType)
		{
			var assembly = typeof(App).GetTypeInfo().Assembly;
			Position position = new Mapsui.UI.Forms.Position(latitude, longitude);
			Pin pin = new Pin(mapView)
			{
				Label = label,
				Address = address,
				Position = position,
				IsVisible = true,
				Type = PinType.Icon,
				Scale = 0.3f,
				Transparency = 0f
			};

			pin.Icon = assembly.GetManifestResourceStream(iconType.ToAssemblyFile()).ToBytes();
			pin.Feature.Geometry = SphericalMercator.FromLonLat(position.Longitude, position.Latitude);
			pin.Feature["name"] = label;

			//hack to ensure that _mapView.CreateCallout doesn't get called because Pin._callout is not null
			FieldInfo privateCalloutProp = pin.GetType().GetField("_callout", BindingFlags.NonPublic | BindingFlags.Instance);
			if (privateCalloutProp != null)
			{
				FieldInfo mapControlProp = mapView.GetType().GetField("_mapControl", BindingFlags.NonPublic | BindingFlags.Instance);
				if (mapControlProp != null)
				{
					Callout callout = new Callout((MapControl)mapControlProp.GetValue(mapView))
					{
						RectRadius = 5,
						ArrowHeight = 20,
						ArrowWidth = 20,
						ArrowAlignment = ArrowAlignment.Top,
						ArrowPosition = 0.5f,
						Title = label,
						Subtitle = address,
						Type = !string.IsNullOrEmpty(address) ? CalloutType.Detail : CalloutType.Single
					};
					if (!string.IsNullOrEmpty(address))
					{
						callout.SubtitleLabel.LineBreakMode = Xamarin.Forms.LineBreakMode.WordWrap;
						callout.SubtitleLabel.HorizontalOptions = LayoutOptions.StartAndExpand;
						callout.SubtitleLabel.FontSize = 12;
					}

					privateCalloutProp.SetValue(pin, callout);
					pin.CalloutAnchor = new Point(0, -20);
				}

			}
			return pin;
		}
`

**Comment by charlenni** [comment]
Thank you for reporting this. I try to look into this.

**Comment by charlenni** [comment]
Ok, I updated the sample (and all other parts) to Xamarin.Forms 3.5.0 and tested it on UWP, but couldn't find any problem.

Could you say, on which platform you have this problem?

**Comment by mmichelbb** [comment]
The issue occurred when debugging with a LG K9 Phone running Android 8.1.
Actually I forgot to mention that I compiled the mapsui lib in debug mode.

**Comment by charlenni** [comment]
Tested it here on my Huawei P8 lite 2017 with Android 7.0 and have no problems. Callouts work like a charme. Than I used the Android emulator on Windows 10 with Android 8.1 and again haven't any problem.

All tests are made with the Forms "Add Pin Sample" sample.

**Comment by mmichelbb** [comment]
Thanks for your efforts! This is quite strange as I encountered the issue constantly when debugging. Every time I accessed Pin.Callout my app stopped responding. So I just compiled the PinSample and ran it on said phone and you might guess: It works. As far as I can see, the pin sample includes it's shared code as shared code, while my app's shared compiles as a .net standard 2.0 class library.

In order to not make you spend more of your time chasing ghosts, I will try to create a basic app which runs into the error and attach it.

**Comment by charlenni** [comment]
That would be perfect. Thank you.

**Comment by mmichelbb** [comment]
So as promised, I put together my version of the Implementation. I kept it as close as possible to the current implementation used in my main app. 

When you compile the app, it will load and display a map of Germany with few markers using the default Callouts, which works fine.

However, as soon as you comment in the lines 99 to 105 in MapsUiHelper.cs located in the shared PinSampleBroken library, the map will never load leaving the app open with a blank screen.
[PinSampleBroken.zip](https://github.com/Mapsui/Mapsui/files/2889928/PinSampleBroken.zip)

**Comment by mmichelbb** [comment]
![infinityloop](https://user-images.githubusercontent.com/24935832/53182975-ced9ff00-35fa-11e9-9ae2-c8deaeb27cd9.png)

**Comment by charlenni** [comment]
I can reproduce the bug and found another one :( Thanky for reporting.

I need to redesign the creatition of callouts. Perhaps I should add them as symbol style to Mpsui, but than it wouldn't possible to use any View of Xamarin.Forms. Have to think about this.

**Comment by lumberjack4** [comment]
I ran into this issue today on UWP.  I seemed to have success reproducing it when running in release mode or when optimizing code debug mode.

A simple, quick fix that worked for me was to change the `private Callout _callout;` field to `volatile`

It might get you by in the short term.

**Comment by mmichelbb** [comment]
> I can reproduce the bug and found another one :( Thanky for reporting.

Does the other bug have anything to do with custom content not showing up in the callout? I encountered this one. I ended up grabbing the parent grid from the exposed labels, removing the initial labels and injecting my UI manually. Not a beauty but works for now.

**Comment by charlenni** [comment]
@mmichelbb You want to add your own view to the callout, right? Do you use it? I, if I abandon this and rewrite the callout, so that it is handled by the Mapsui part and not the Mapsui.Forms part. With this in mind, custom content will be not possible anymore.

**Comment by mmichelbb** [comment]
I tried to put a Xamarin.Forms ContentView into the Content Property of the Callout. However, i have never managed to make it appear on screen, the callout was always empty. But anyways, I guess it doesn't make much sense putting a lot of debugging time into this, since it will probably change soon and in the meanwhile i have a working solution.

**Comment by charlenni** [comment]
This doesn't answer my question. Do you use own content for the callout?

**Comment by mmichelbb** [comment]
Yes, but not in the first place. The initial infinity loop bug appeared as I had my code like the example I attached here. After I got that running using the reflection injection method to create the callout, I tried setting Pin.Content to a custom content view. 

As this didn't work for me, since I always ended up in looking at the callout box with nothing but the close button in it, I did this after setting Pin._callout by reflection:

`
Grid calloutGrid = callout.TitleLabel.Parent as Grid;
					if (calloutGrid != null)
					{
						calloutGrid.Children.Remove(callout.TitleLabel);
						calloutGrid.BackgroundColor = darkMode ? (Color)(App.Current as App).Resources["ModernCalendarEntryBgDark"] : (Color)(App.Current as App).Resources["ModernCalendarEntryBgLight"];
						

						CustomFontLabel titlelabel = new CustomFontLabel();
						titlelabel.FontFamily = (string)(App.Current as App).Resources["DefaultFontHeadline"];
						titlelabel.FontSize = 16;
						titlelabel.VerticalOptions = LayoutOptions.Center;
						titlelabel.Text = label;
						titlelabel.Margin = new Thickness(18, 0, 0, 0);
						titlelabel.TextColor = statusColor != null ? statusColor.Value : Color.Gray;
						
						calloutGrid.Children.Add(titlelabel, 0, 0);

						if (!string.IsNullOrEmpty(address))
						{
							
							BoxView bgbox = new BoxView();
							bgbox.BackgroundColor = darkMode ? (Color)(App.Current as App).Resources["ModernCalendarEntryOverlayDark"] : (Color)(App.Current as App).Resources["ModernCalendarEntryOverlayLight"];
							Grid.SetColumn(bgbox, 0);
							Grid.SetRow(bgbox, 1);
							Grid.SetColumnSpan(bgbox, 2);
							calloutGrid.Children.Add(bgbox);

							CustomFontLabel subtitlelabel = new CustomFontLabel();
							subtitlelabel.Text = address;
							subtitlelabel.FontFamily = (string)(App.Current as App).Resources["DefaultFontLight"];
							subtitlelabel.FontSize = 12;
							subtitlelabel.Margin = new Thickness(18, 0, 0, 2);
							subtitlelabel.TextColor = darkMode ? Color.White : Color.Black;
							subtitlelabel.LineBreakMode = Xamarin.Forms.LineBreakMode.WordWrap;
							subtitlelabel.HorizontalOptions = LayoutOptions.StartAndExpand;
							subtitlelabel.VerticalOptions = LayoutOptions.StartAndExpand;
							calloutGrid.Children.Remove(callout.SubtitleLabel);
							calloutGrid.Children.Add(subtitlelabel, 0, 1);
						}

						BoxView box = new BoxView();
						box.BackgroundColor = statusColor != null ? statusColor.Value : Color.LightGray;
						box.WidthRequest = 12;
						box.Margin = new Thickness(0, 0, 2, 2);
						box.HorizontalOptions = LayoutOptions.Start;
						Grid.SetRow(box, 0);
						Grid.SetColumn(box, 0);
						Grid.SetRowSpan(box, 2);

						calloutGrid.Children.Add(box);
						// Remove and add closeButton again to make it appear in front of other elements
						Button closeButton = calloutGrid.Children.OfType<Button>().FirstOrDefault();
						if (closeButton != null)
						{
							calloutGrid.Children.Remove(closeButton);
							calloutGrid.Children.Add(closeButton);
						}
					}
`
which works fine for now.

**Comment by pauldendulk** [comment]
@charlenni I am not familiar with the code of this part. How serious is this? Does it need to be fixed before we release a 'release candidate'? I looks like most problems are solved now, so I was thinking of going for a release candidate soon.

**Comment by charlenni** [comment]
@pauldendulk I'm not very happy with the callout implementation. There are problems. First of all is the afore mentioned one. It seems to happen, because Xamarin.Forms are not ready with initialising the views of the page when accessing the callouts. The next is the lag of callout movement when moving the map fast around. The callout has than no conntact to the pin it belongs to. The only advantage is, that you could add each view you want.

When we decide to have only a few default layouts for callouts, it would be possible to transfer the rendering to the normal Skia renderer and draw it like all other things on the map. Perhaps as a style, which takes things like title or subtitle from _dictionary of feature. Lets say, we could have a title, title + subtitle, both with a icon (left or right side) and a bitmap style. With the last one, it would be possible to draw user designed callouts.

Do you think, this would be possible?

**Comment by pauldendulk** [comment]
'like a style', so the callout would be somewhat like a label? 

I think you have to make a lot of specific choices to implement that. And many users will want something that is just slightly different. 

Custom renderers for styles might also be an option. This is a big change for Mapsui but will offer a lot of possibilities. I think many users could create nice things with their own skia rendering.

**Comment by pauldendulk** [comment]
On the other hand. If we implement more in the skia renderer we do not have to implement it for every platform. Much of what is now in MapView will also be requested for other platforms. We should not try to implement that in every platform separately.

**Comment by charlenni** [comment]
Perhaps SKPicture or SKDrawable is a way. Both are recorders for Skia drawing commands. They could be implemented very easy and are very flexible. While rendering, they are drawn like SKImages.

**Comment by pauldendulk** [comment]
A custom renderer is for another version. What can we do for the release? Users do not really need the MapView callout, do they? They could also add their own?

**Comment by mmichelbb** [comment]
@pauldendulk As long as you don't add a custom callout, you are fine, even when using MapsUI in a .Net Standard Class library.
However as soon as you reference the callout, you run into the mentioned infinity loop. This does not only happen with custom callouts, but also with the default callout provided by MapsUI.

**Comment by charlenni** [comment]
We could remove it. It was a try. 

@mmichelbb The problem takes only placde when you acces the callout while creating the MapView. If you access the callout after the MapView is created, you don't have any problems.

**Comment by mmichelbb** [comment]
@charlenni I may have found a solution for it. I replaced that dubious while loop in MapView.cs with a thread safe sync lock. It works for both, the original AddPin Sample (Shared Code) as well as the implementation I made (.Net Standard lib). I am now rebuilding the library so that I can include it in my app to check if it works with iOS, too.

See https://github.com/mmichelbb/Mapsui/commit/04beb7bbb327e232b699b1166df983cc5c249471 for more details.

**Comment by pauldendulk** [comment]
How about using 'native' Mapsui pins and using a Xamarin.Forms callout that is positioned correctly on the location?

**Comment by charlenni** [comment]
That's the way it is working ;)

**Comment by charlenni** [comment]
I removed the BeginInvoke onMainThread and tested it on Android, where it runs perfect. I'm not sure, if we need to create the callout in the main thread. It isn't displayed at this time. Had to try, if this works on iOS too.

**Comment by pauldendulk** [comment]
_mapView.CreateCallout is called from the Callout field of Pin. Perhaps it is better to let the MapView create the Callout on startup, but make it invisible. Make it visible and position it on the MapInfo event within the MapView. This way only the MapView needs access to the Callout. In general limiting the number of pointers is a good thing.

**Comment by pauldendulk** [comment]
![image](https://user-images.githubusercontent.com/963462/54222232-49f54d80-44f5-11e9-8c9b-46767571983c.png)

**Comment by charlenni** [comment]
But if you have hundreds of pins, you will create hundrets of callouts at startup. This is a waste of resource because you use none of them.

**Comment by pauldendulk** [comment]
There is just one callout. You only show it for the feature clicked.

**Comment by charlenni** [comment]
No, it depends on the flag Pin.Callout.IsClosableByClick. If this is true, the callout is closed, if you click anywhere outside of the callout. If you set this flag to false, you could have as much callouts as pins (and even more ;) ).

**Comment by charlenni** [comment]
![grafik](https://user-images.githubusercontent.com/4381223/54226002-29c98c80-44fd-11e9-8b60-f26918f7fec8.png)

**Comment by klzig** [comment]
FWIW it would not break my heart if you restricted callouts to display one at a time, for the selected pin.  My app has no use for multiple callouts being displayed simultaneously.

**Comment by pauldendulk** [comment]
@charlenni ah, I see, nice functionality but perhaps this causes too much complications at this point. It could also cause problems with performance because we don't know how many points there are and it might be complicated to clean the callouts up when points are removed or the whole layer is removed. 

The 'one callout' alternative is perhaps not very hard to implement given the current solution. 

Showing a callout for each pin could perhaps could be implemented with labels. The LabelSymbol has an Enabled state to turn them on or off. Current labels are very simple but they were always intended as temporary solutions to be replaced with labels that could be properly styled. 

In this case a custom label renderer might also be a good solution for many users. They could do their own skia rendering to draw their own callouts. This could also stimulate contributions.

**Comment by pauldendulk** [comment]
@charlenni. The 'native callout' is now discussed in another issue. This is change that will take some time to realize. For the short term my focus is to release a RC version of 2.0. So we should work on some solution for the crash. The 'single-callout' solution might also be feasible for the short term, but just removing the Pin layer would be an option. What do you think?

**Comment by charlenni** [comment]
Removing the pin layer removes the main reason for MapView.

Perhaps it is enough to remove the BeginInvokeOnMainThread. But I have to test this.

**Comment by charlenni** [comment]
If you like to publish a new version, than remove the Forms part.

**Comment by lumberjack4** [comment]
Putting the callout creation on the background thread for UWP throws a cross thread exception on beta 25.  Can be seen when running the UWP Pin sample on pin creation.

**Comment by pauldendulk** [comment]
@lumberjack4 I can reproduce this.

**Comment by lumberjack4** [comment]
Putting the callout creation back on the UI thread and making the `_calout` field `volatile` works for me in UWP.  Without volatile I can see that the callout is always successfully created when debugging but it appears that since the while loop never sees that the callout has been created.  Making the field volatile appears to force the loop to refresh the value and exit.  My code is below

```
        private volatile Callout _callout;

        /// <summary>
        /// Creates a callout at the given position
        /// </summary>
        /// <returns>The callout</returns>
        /// <param name="position">Position of callout</param>
        public Callout CreateCallout(Position position)
        {
            Device.BeginInvokeOnMainThread(() => _callout = new Callout(_mapControl) { Anchor = position });

            // My interpretation (PDD): This while keeps looping until the asynchronous call
            // above has created a callout.
            // An alternative might be to avoid CreateCallout from a non-ui thread by throwing
            // early.
            while (_callout == null) ;

            var result = _callout;
            _callout = null;

            return result;
        }
```

**Comment by jasells** [comment]
I see this is closed, but no releases since before this thread.  Are there plans to address?

**Comment by jasells** [comment]
Ok, I can confirm that this is not resolved in current master for UWP (most likely also WPF).  

1. An error: " The application called an interface that was marshalled for a different thread " is thrown in the MapsUI.Samples.Forms sample when placing pins on the map by clicking the map.
2. Even with lumberjack's nominal fix (of the current sample), there is still another use-case broken: if CreateCallout() is called from the UI thread, it will "deadlock" (most times, but possibly not always) as it is the UI thread in both the `while (_callout == null)` and in `Device.BeginInvokeOnMainThread(() => _callout = new Callout(_mapControl)`...  so a thread waiting on itself will never boil... I mean complete.

See [here ](https://github.com/jasells/Mapsui/tree/callout-bug-UWP-repro)for case 1 (current master) repro (may require optimized code option in build settings of debug, or change to release)

The root of the issue is that the MapControl.Touch event (from SKGLView) is raised from a worker thread, and the ctor for Callout is creating and setting Grids and other UI elements that don't like being created and/or set from non-UI threads.  This is a long time idiom in Windows, even back to Windows Forms and .NET 1.x (I know this monster well... we have history).  The fact that the Touch event is fired from a non-UI thread is a little surprising, as I thought these type of events were normally marshalled back to the UI thread before firing so as to avoid this exact problem.  Maybe this non-intuitive behavior is a Skia thing?

See [here ](https://github.com/jasells/Mapsui/tree/callout-bug)for a working fix, that affects no platforms other than UWP and WPF.  I don't know if other platforms need the fix... but this fix is easily adjusted to any or all platforms in need, since it is only an issue in the Mapsui.UI.Forms code base, and not in the "native" platform MapsUI code base(s).

My experience with Xamarin Forms on Android is that most of this type of context switching is baked in already either internally to Xamarin.Forms, or into the native Android UI elements (controls) to avoid these issues like case 1.  However, it is easy to introduce a new issue on Android with an attempted fix that may seem to work in Windows for this specific case but causes a spin-wait or deadlock.  

I marshalled all the Touch events back to the UI thread in [ Mapsui.UI.Forms/MapView.cs](https://github.com/jasells/Mapsui/blob/1c63d5e152aba444c6a7b191232d50d87dc924e1/Mapsui.UI.Forms/MapView.cs#L70) partially shown: 
```C#
    EventHandler<TappedEventArgs> handler = (s, e) =>
    {
        if (PlatformRequiresUIcontext())
        {    
            Device.BeginInvokeOnMainThread(() => HandlerTap(s, e));
        }
        else
        {
            HandlerTap(s, e);
        }
    };

    _mapControl.SingleTap += handler;
    _mapControl.DoubleTap += handler;
```
